### PR TITLE
feat(mv3-part-2): Browser Adapter Event Manager

### PR DESCRIPTION
**Pull request by madalynrose**
#### Details

This PR creates:
 * `SetTimeoutManager`: class that calls either `windowUtils.setTimeout` or `alarmUtils.createAlarmWithCallback` depending on what it is passed. This will be used when instantiating `TabContextFactory` and `InjectionController` in init scripts.
 * `BrowserAdapterEventManager`: will allow the browser adapter used in background scripts to manage its own events by:
   * maintaining an internal list of application-level listeners
   * synchronously registering an adapter-level listener that calls registered application-level listeners for the given event.
   *  deferring events for which a listener is not found until one is registered
   *  preventing watchdog timeouts by wrapping listener calls in a 4-minute promise race
   * ensuring suspected asynchronous calls have time to finish by setting a 2-minute timer and waiting when an async response is received

##### Motivation

feature work

##### Context

Incorporates work from #5366.  A later PR will use these classes inside of the `BrowserAdapter` and  `TabContextFactory` used in the service worker.

#### Pull request checklist

- [n/a] Addresses an existing issue: #0000
- [x] Ran `yarn fastpass`
- [x] Added/updated relevant unit test(s) (and ran `yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] PR title *AND* final merge commit title both start with a semantic tag (`fix:`, `chore:`, `feat(feature-name):`, `refactor:`). See `CONTRIBUTING.md`.

### CODE REVIEW

File: src/background/set-timeout-manager.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+export class SetTimeoutManager {
+    constructor(private utils: { alarmUtils?: AlarmUtils; windowUtils?: WindowUtils }) {}
+
+    public setTimeout(callback: Function, delayInMs: number, name?: string): void {
+        if (name && this.utils.alarmUtils) {
+            this.utils.alarmUtils.createAlarmWithCallback(name, delayInMs, callback);
+        } else {
+            this.utils.windowUtils?.setTimeout(callback, delayInMs);
+        }
```
Comment (sfoslund) [inline src/background/set-timeout-manager.ts:15]: nit: Should we check that windowUtils is not null, just in case? Maybe add the check and log an error when neither the alarm or window utils are non-null?

File: src/tests/unit/tests/background/set-timeout-manager.test.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import { WindowUtils } from 'common/window-utils';
+import { IMock, Mock, Times } from 'typemoq';
+
+describe(`setTimeout`, () => {
+    const testTimeoutId = `timeout-promise-${Date.now()}`;
+    const delayInMs = 50;
+    const callback = () => true;
+
+    describe('SetTimeoutManager', () => {
+        describe('with alarmUtils', () => {
```
Comment (sfoslund) [inline src/tests/unit/tests/background/set-timeout-manager.test.ts:14]: Can we add some error test cases here, for example, when both alarm and window utils are null, or when name and window utils are both null?

File: src/common/browser-adapters/browser-adapter-event-manager.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    private async forwardEventToApplicationListener(
+        listener: ApplicationListener,
+        eventArgs: any[],
+    ): Promise<any> {
+        const result = listener(...eventArgs);
+        if (!!result && typeof result.then === 'function') {
+            if (this.promiseFactory) {
+                //wrapping application listener promise responses in a 4-minute promise race
+                //prevents the service worker going idle before a response is sent
+                await this.promiseFactory.timeout(result, 240000);
```
Comment (sfoslund) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:68]: nit: Can we save 240000 as a constant instead of using it as a magic number? (Same with 12000 below)

File: src/common/browser-adapters/browser-adapter-event-manager.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            }
+        } else {
+            // it is possible that this is the result of a fire and forget listener
+            // wrap in 2-minute timeout to ensure it completes during service worker lifetime
+            if (this.setTimeoutManager) {
```
Comment (sfoslund) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:73]: Is there ever a case when we reasonably expect the setTimeoutManager to be null? If not, can we do this check in the constructor or make it non-nullable?

File: src/common/browser-adapters/browser-adapter-event-manager.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        const allListeners = this.eventsToApplicationListenersMapping[eventType];
+        const applicationListenerIndex = allListeners.findIndex(listener);
+        if (applicationListenerIndex < 0) {
+            console.error('no application listeners found for ', eventType);
+        } else {
+            if (allListeners.length === 1) {
+                delete this.eventsToApplicationListenersMapping[eventType];
+            } else {
+                this.eventsToApplicationListenersMapping[eventType] =
+                    allListeners.splice(applicationListenerIndex);
```
Comment (sfoslund) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:101]: I'm confused by this logic- if the listener we want to remove is in the middle of allListeners, won't this remove the following items in allListeners as well?

File: src/common/browser-adapters/browser-adapter-event-manager.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            }
+        }
+    }
+
+    public removeListener(
```
Comment (sfoslund) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:106]: Is there any test coverage for this method?

**Review by sfoslund — COMMENTED** [review]
Several questions/ comments

File: src/common/browser-adapters/browser-adapter-event-manager.ts (hunk 6 above)
Comment (madalynrose) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:106]: There isn't right now because it isn't used anywhere. None of the unregister or remove methods are. I'd plan on adding tests once we determine what graceful teardown looks like.

File: src/common/browser-adapters/browser-adapter-event-manager.ts (hunk 5 above)
Comment (madalynrose) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:101]: oh my you're right! I need to add the argument for how many to remove (just the one!).

File: src/common/browser-adapters/browser-adapter-event-manager.ts (hunk 4 above)
Comment (madalynrose) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:73]: good point. this depends on implementation details for when we actually use this inside browser adapters and really is out of scope for this PR. Going to make them required and get rid of this logic.

File: src/common/browser-adapters/browser-adapter-event-manager.ts (hunk 5 above)
Comment (madalynrose) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:101]: I ended up including the remove function from lodash to make this cleaner

File: src/common/browser-adapters/browser-adapter-event-manager.ts (hunk 6 above)
Comment (sfoslund) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:104]: Makes sense, yeah I'd prefer to omit them for now just to make sure we don't add them now then forget to add tests later

File: src/background/set-timeout-manager.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { AlarmUtils } from 'background/alarm-utils';
+import { WindowUtils } from 'common/window-utils';
+
+export class SetTimeoutManager {
+    constructor(private utils: { alarmUtils?: AlarmUtils; windowUtils?: WindowUtils }) {}
```
Comment (pownkel) [inline src/background/set-timeout-manager.ts:8]: It's a little unclear to me when/why AlarmUtils or WindowUtils would be undefined. Is WindowUtils a temporary measure during the mv3 transition, or will we need to use this class in a context where window is defined? If we need both paths long-term, could this class be separated into something like BackgroundSetTimeoutManager and BrowserSetTimeoutManager?

Comment (madalynrose) [reply src/background/set-timeout-manager.ts:8]: Oh that's a good point. I'll split this in two similar to how the `PromiseFactory` stuff is structured.

File: src/common/browser-adapters/browser-adapter-event-manager.ts (hunk 6 above)
Comment (madalynrose) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:120]: I ended up writing tests for them after remembering where they are used later. We'll need them for my next PR anyhow.

File: src/common/timeouts/timeout-factory.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+export enum TimeoutType {
+    Window,
+    Alarm,
+}
+export interface TimeoutFactory {
+    timeoutType: TimeoutType;
+    createTimeout: Timeout;
+}
+
+export class WindowTimeoutFactory implements TimeoutFactory {
```
Comment (sfoslund) [inline src/common/timeouts/timeout-factory.ts:17]: This structure looks much better, great suggestion @pownkel!

**Review by sfoslund — APPROVED** [review]
LGTM!

File: src/common/browser-adapters/browser-adapter-event-manager.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        this.deferredEvents = this.deferredEvents.filter(
+            event => this.processEvent(event.eventType, event.eventArgs, true) === false,
+        );
+    }
+
+    private async forwardEventToApplicationListener(
+        listener: ApplicationListener,
+        eventArgs: any[],
+    ): Promise<any> {
+        const result = listener(...eventArgs);
```
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:63]: This should probably be defensive about the `listener` throwing

File: src/common/browser-adapters/browser-adapter-event-manager.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+const TWO_MINUTES = 120000;
+const FOUR_MINUTES = 2 * TWO_MINUTES;
+
+export class BrowserAdapterEventManager {
+    protected deferredEvents: EventDetails[] = [];
+    protected eventsToApplicationListenersMapping: DictionaryStringTo<ApplicationListener[]> = {};
+    protected adapterListener: AdapterListener =
+        (eventType: string) =>
+        (...eventArgs: any[]) => {
+            this.processEvent(eventType, eventArgs);
```
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:30]: If this is intended to be the thing that gets attached to browser events directly, it needs to be returning the promise generated by `forwardEventToApplicationListener`, no?

Comment (dbjorge) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:30]: (and probably creating some sort of deferred-with-timeout Promise in the event that there are no current app listeners)
